docs/developer-guide/partition-data.rst

```diff
@@ -10,26 +10,69 @@ of the join have the same configuration settings for partitions.
 
-.. _keys-have-same-keying-scheme:
+.. _keys-have-same-key-schema:
 
-Records Have the Same Keying Scheme
-===================================
+Records have the Same Key Schema
+================================
 
-The input records for the join must have the same keying scheme, which means
-that the join must use the same key field on both sides.
+For a join to work the keys from both sides must have the same serialized binary data.
 
-For example, you can join a stream of user clicks that's keyed by a ``VARCHAR userId``
-field with a table of user profiles that's keyed by a ``VARCHAR userId`` field. 
-The join won't match if the key fields don't have the same name and type.
+For example, you can join a stream of user clicks that's keyed on a ``VARCHAR`` user id with a table
+of user's profiles that's also keyed on a ``VARCHAR`` user id. Records with the exact same user id
+on both sides will be joined.
+
+Tables created on top of existing Kafka topics, for example those created with a ``CREATE TABLE``
+statement, will be keyed on the data held in the key of the records in the
+Kafka topic.  KSQL presents this data in the ``ROWKEY`` column, and expects the data to be a ``VARCHAR``.
+
+Tables created inside KSQL from other sources, for example those created with a ``CREATE TABLE AS SELECT``
+statement, will copy the key from their source(s) unless there is an explicit ``GROUP BY`` or ``PARTITION BY``
+clause, which can changed what the table is keyed on.
+
+KSQL will automatically repartition a stream of data before joining, if required. In the example below
+KSQL is used to join a stream of clicks with a table of users.  The ``clicks`` stream is not keyed on anything
+initially. The ``users`` table keyed on ``userId``.  Because the stream is being joined on a column other than
+it's key field, i.e. ``ROWKEY``, KSQL will automatically repartition the stream under the hood.
+
+.. code:: sql
+
+   -- clicks stream, with an unknown key.
+   CREATE STREAM clicks (userId STRING, url STRING) WITH(kafka_topic='clickstream', value_format='json');
+
+   -- users table, where the records have the userId in the record's key:
+   CREATE TABLE  users  (userId STRING, fullName STRING) WITH(kafka_topic='users', value_format='json', key='userId');
+
+   -- join will automatically repartition clicks stream:
+   SELECT clicks.url, users.fullName FROM clicks JOIN users ON clicks.userId = users.userId;
+
+.. note::
+
+   While KSQL will automatically repartition a stream should a join require it, KSQL will reject any join
+   any table's column that is not the key. This is because KSQL does not support joins on foreign keys
+   and repartitioning a table would corrupt the data.
+
+If you are using the same sources in more than one join that requires the data to be repartitioned you
+may choose to repartition manually to avoid KSQL repartitioning multiple times.
+
+To repartition a stream, use the PARTITION BY clause.

```

Comment:
thinking about it, should there be caution here that repartitioning the stream will change the ordering of the stream's events?

Reply:
Yep - that was in my head to do, must of forgotten.

Reply:
> not sure that's strictly correct ? sure, the records from your original partitions may now be interleaved together in a new partition but their relative ordering wrt each other is unchanged. given that the basic ordering guarantee of Kafka only applies within a partition anyway, that record-relative ordering guarantee is unbroken even if you reshuffle those same messages into new partitions, no?
looking at it more, perhaps its better to say that "the ordering of the stream's events" is what is incorrect here - there is no such guarantee in the first place....

Reply:
classic case of "if you don't like the answer, then the question must be wrong" ? :-)

Reply:
Updated - what do you think?

Reply:
What do you think of:

> Be aware that Kafka guarantees the relative order of any two messages from one source partition only if they are also both in the same partition after the repartition. Otherwise, Kafka is likely to interleave messages.